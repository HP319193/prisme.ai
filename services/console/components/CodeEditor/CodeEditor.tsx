import { forwardRef, useEffect, useImperativeHandle, useRef } from "react";
import AceEditor, { IAceEditorProps, IMarker } from "react-ace";
import ReactAce from "react-ace/lib/ace";
import "ace-builds/src-noconflict/ext-language_tools";
import "ace-builds/src-noconflict/ext-searchbox";
import "ace-builds/src-noconflict/snippets/javascript";
import "ace-builds/src-noconflict/theme-xcode";
import "ace-builds/src-noconflict/mode-javascript";
import "ace-builds/src-noconflict/mode-css";
import "ace-builds/src-noconflict/mode-html";
import "ace-builds/src-noconflict/mode-yaml";
import "ace-builds/src-noconflict/mode-json";

import { Ace } from "ace-builds";
import { addCustomAnnotations } from "../../utils/aceEditorTweaks";

export interface CodeEditorProps extends IAceEditorProps {
  mode: "javascript" | "css" | "html" | "yaml" | "json";
  value: string;
  onChange?: (value: string) => void;
  placeholder?: string;
  height?: string;
  width?: string;
  readOnly?: boolean;
  completers?: Ace.Completer[];
  annotations?: Ace.Annotation[];
  markers?: IMarker[];
  shortcuts?: {
    name: string;
    bindKey: {
      mac: string;
      win: string;
    };
    exec: () => void;
  }[];
  scrollTo?: [number, number];
  enableAutocompletion?: boolean;
  maxLines?: number;
}

export const CodeEditor = forwardRef<AceEditor, CodeEditorProps>(
  function CodeEditor(
    {
      value: initialValue,
      onChange,
      placeholder = "",
      mode,
      height = "auto",
      width = "auto",
      readOnly = false,
      completers,
      shortcuts,
      annotations,
      markers,
      scrollTo,
      style,
      ...props
    }: CodeEditorProps,
    ref
  ) {
    const value = useRef(initialValue);
    const aceRef = useRef<ReactAce>(null);
    useImperativeHandle(ref, () => aceRef.current!, []);
    //const theme = `solarized_${darkMode ? "dark" : "light"}`;
    const theme = `xcode`;

    useEffect(() => {
      if (!aceRef.current) return;
      addCustomAnnotations(aceRef.current.editor, annotations);
    }, [annotations]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;

      (completers || []).forEach((completer) =>
        editor.completers.push(completer)
      );
      // This code automatically opens autocomplete menu when typing `.` character
      editor.commands.on("afterExec", (e) => {
        if (e.command.name === "insertstring" && /^[\w.]$/.test(`${e.args}`)) {
          if (!aceRef.current) return;
          aceRef.current.editor.execCommand("startAutocomplete");
        }
      });

      return () => {
        (completers || []).forEach((completer) => {
          const pos = editor.completers.indexOf(completer);
          if (pos === -1) return;
          editor.completers.splice(pos, 1);
        });
      };
    }, [completers]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;
      (shortcuts || []).forEach((command) => {
        editor.commands.addCommand(command);
      });
    }, [shortcuts]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;
      Object.keys(editor.getSession().getMarkers()).forEach((key) => {
        editor.getSession().removeMarker(+key);
      });
      (markers || []).forEach((marker) => {
        const { className, endCol, endRow, startCol, startRow, type } = marker;
        const Range = ace.require("ace/range").Range;
        editor
          .getSession()
          .addMarker(
            new Range(startRow, startCol, endRow, endCol),
            className,
            type
          );
      });
    }, [markers]);

    return (
      <AceEditor
        ref={aceRef}
        style={{
          display: "flex",
          flex: 1,
          ...style,
        }}
        mode={mode}
        theme={theme}
        onChange={(v: string) => {
          value.current = v;
          onChange && onChange(v);
        }}
        value={value.current}
        enableBasicAutocompletion={true}
        enableLiveAutocompletion={true}
        enableSnippets={true}
        readOnly={readOnly}
        tabSize={2}
        placeholder={placeholder}
        showPrintMargin={true}
        showGutter={true}
        highlightActiveLine={true}
        height={height}
        width={width}
        annotations={annotations}
        {...props}
      />
    );
  }
);

export default CodeEditor;
