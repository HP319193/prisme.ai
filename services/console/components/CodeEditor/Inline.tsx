import * as React from 'react';

import CodeEditor, { CodeEditorProps } from './CodeEditor';

export const CodeEditorInline: React.FC<CodeEditorProps> = ({
  value,
  className,
  ...props
}) => {
  const singleLine = `${value}`.split(/\n/).length === 1;

  return (
    <React.Suspense fallback={<div />}>
      <CodeEditor
        {...props}
        style={{
          flex: 1,
          minHeight: '50px',
          border: '1px solid #E5E5E5',
          borderRadius: '0.3rem',
          padding: '10px',
          ...(singleLine ? undefined : { background: 'white' }),
          ...props.style,
        }}
        value={
          typeof value === 'string' ? value : JSON.stringify(value, null, '  ')
        }
        maxLines={Infinity}
        showGutter={!singleLine}
        className={`${value ? 'filled' : ''} ${className}`}
      />
    </React.Suspense>
  );
};

export default CodeEditorInline;
