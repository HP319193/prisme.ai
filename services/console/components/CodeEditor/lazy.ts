import dynamic from "next/dynamic";

export const CodeEditor = dynamic(import("./"), { ssr: false });

export default CodeEditor
