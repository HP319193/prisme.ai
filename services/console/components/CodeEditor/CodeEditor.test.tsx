import AceEditor, { IMarker } from "react-ace";
import CodeEditor from "./CodeEditor";
import renderer, { act } from "react-test-renderer";
import { Ace } from "ace-builds";
import { useRef } from "react";

it("should render", () => {
  const root = renderer.create(<CodeEditor mode="yaml" value="foo" />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should update value", () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <CodeEditor mode="yaml" value="foo" onChange={onChange} />
  );
  act(() => {
    root.root.findByType(AceEditor).props.onChange("bar");
  });
  expect(onChange).toHaveBeenCalledWith("bar");
});

it("should set completers", async () => {
  const completers: Ace.Completer[] = [
    {
      getCompletions: () => null,
    },
  ];
  let ref: any;
  const Container = () => {
    ref = useRef<AceEditor>(null);
    return (
      <CodeEditor ref={ref} mode="yaml" value="foo" completers={completers} />
    );
  };
  const root = renderer.create(<Container />);
  await act(async () => {
    await true;
  });
  const editor = ref.current.editor;
  expect(editor.completers).toContain(completers[0]);
  act(() => {
    root.unmount();
  });
  expect(editor.completers).not.toContain(completers[0]);
});

it("should have an auto completers on .", async () => {
  let ref: any;
  const Container = () => {
    ref = useRef<AceEditor>(null);
    return <CodeEditor ref={ref} mode="yaml" value="foo" />;
  };
  const root = renderer.create(<Container />);
  await act(async () => {
    await true;
  });
  const editor = ref.current.editor;
  jest.spyOn(editor, "execCommand");
  editor.execCommand("insertstring", ".");
  expect(editor.execCommand).toHaveBeenCalledWith("startAutocomplete");
});

it("should set shortcuts", async () => {
  const shortcuts: any[] = [
    {
      name: "foo",
    },
  ];
  let ref: any;
  const Container = ({ shortcuts }: any) => {
    ref = useRef<AceEditor>(null);
    return (
      <CodeEditor ref={ref} mode="yaml" value="foo" shortcuts={shortcuts} />
    );
  };
  const root = renderer.create(<Container />);
  await act(async () => {
    await true;
  });
  const editor = ref.current.editor;
  jest.spyOn(editor.commands, "addCommand");
  act(() => {
    root.update(<Container shortcuts={shortcuts} />);
  });
  expect(editor.commands.addCommand).toHaveBeenCalledWith({
    name: "foo",
  });
  editor.commands.addCommand.mockClear();
  act(() => {
    root.update(
      <Container
        shortcuts={[
          {
            name: "foo",
            bar: "foo",
          },
        ]}
      />
    );
  });
  expect(editor.commands.addCommand).not.toHaveBeenCalled();
});

it("should set markers", async () => {
  const markers: IMarker[] = [
    {
      className: "foo",
      endCol: 1,
      endRow: 1,
      startCol: 1,
      startRow: 1,
      type: "fullLine",
    },
  ];
  let ref: any;
  const Container = ({ markers }: any) => {
    ref = useRef<AceEditor>(null);
    return <CodeEditor ref={ref} mode="yaml" value="foo" markers={markers} />;
  };
  const root = renderer.create(<Container />);
  await act(async () => {
    await true;
  });
  const editor = ref.current.editor;
  jest.spyOn(editor.getSession(), "addMarker");
  act(() => {
    root.update(<Container markers={markers} />);
  });
  expect(editor.getSession().addMarker).toHaveBeenCalledWith(
    {
      end: {
        column: 1,
        row: 1,
      },
      start: {
        column: 1,
        row: 1,
      },
    },
    "foo",
    "fullLine"
  );
});
