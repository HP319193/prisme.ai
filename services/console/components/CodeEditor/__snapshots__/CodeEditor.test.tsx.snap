// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  id="ace-editor"
  style={
    Object {
      "display": "flex",
      "flex": 1,
      "height": "auto",
      "width": "auto",
    }
  }
/>
`;
